### ISSUE TITLE
[Squiz.Arrays.ArrayDeclaration.MultiLineNotAllowed] Not reporting single possible arrays

### ISSUE THREAD

**Issue by carusogabriel** [squizlabs/PHP_CodeSniffer#2939]
Hello there 👋 

During the development of https://github.com/doctrine/coding-standard/pull/173, I've notice that the `Squiz.Arrays.ArrayDeclaration` sniff with the following configuration:

```xml
    <!-- Force array declaration structure -->
    <rule ref="Squiz.Arrays.ArrayDeclaration">
        <!-- Disable arrow alignment -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned"/>
        <!-- Uses indentation of only single space -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.KeyNotAligned"/>
        <!-- Allow multiple values on a single line -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed"/>
        <!-- Disable alignment of braces -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned"/>
        <!-- Disable alignment of values with opening brace -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.ValueNotAligned"/>
        <!-- Checked by SlevomatCodingStandard.Arrays.TrailingArrayComma.MissingTrailingComma -->
        <exclude name="Squiz.Arrays.ArrayDeclaration.NoCommaAfterLast"/>
        <exclude name="Squiz.Arrays.ArrayDeclaration.NoComma"/>
    </rule>
```

Didn't catch the following case:

```php
<?php

MyClass::baz([
    new MyObject(),
]);
```

Which I'd expect:

```
FILE: tests/input/array.php
-------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-------------------------------------------------------------------------------------------------
 4 | ERROR | [x] Multi-line array contains a single value; use single-line array instead (Squiz.Arrays.ArrayDeclaration.MultiLineNotAllowed)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATION AUTOMATICALLY
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Is this a bug or something that wasn't configured correctly?

**Comment by jrfnl** [comment]
Hi @carusogabriel That sniff is badly designed which is the cause of most problems with it - see #582 for a far more detailed discussion on it.

The short of it is that when excluding certain error codes, not just that error code, but a lot more of the sniff gets disabled.

**Comment by carusogabriel** [comment]
Hello @jrfnl, thanks for your answer and the linked pull request.

I'll keep an eye on that and once you come up with a solution, I'll test with this issue and close!

**Comment by gsherwood** [comment]
> The short of it is that when excluding certain error codes, not just that error code, but a lot more of the sniff gets disabled.

That's true, but not the problem in this case.

The sniff only forces a single-item array to be defined on the same line if the value is "simple". I simple value consists of one token, like a string, or a boolean, or a number. The statement `new MyObject()` is not simple enough for it to force the value onto one line.

Changing the behaviour is very easy, but a big backwards compatibility break, and will lead to some very messy arrays. Consider this case from the unit tests:
```
$array = array(
    'name'       => 'contactSubject',
    'required'   => TRUE,
    'validators' => array(
        new \Zend\Validator\InArray(array('haystack' => array_keys($aSubjects)))
    ),
);
```
It's probably more readable like this:
```
$array = array(
    'name'       => 'contactSubject',
    'required'   => TRUE,
    'validators' => array(
        new \Zend\Validator\InArray(
            array(
                'haystack' => array_keys($aSubjects)
            )
        ),
    ),
);
```
But would instead require this in all cases:
```
$array = array(
    'name'       => 'contactSubject',
    'required'   => TRUE,
    'validators' => array(new \Zend\Validator\InArray(array('haystack' => array_keys($aSubjects)))),
);
```
For this reason, I'm not going to change the behaviour of the current sniff.

Does that provide enough info to explain things?

**Comment by carusogabriel** [comment]
> Does that provide enough info to explain things?

Yes, it does, thanks for the extensive explanation about the BC Break 👍
